### ISSUE TITLE
Timeout Errors On Large Sites

### ISSUE THREAD

**Issue by HOFB2015** [getgrav/grav#201]
I have a larger site then most people would (95% of pages are static with no further amends). As the site grew from 300-400 pages to 2500+ overnight I am getting time out errors (30 secs). I reload the page perhaps 3-4 times and then the page loads. I move onto another page I get the same issue. Each page I move onto also time outs for several attempts before loading.

Hopefully theirs a large sites setting in Grav :dancers:

**Comment by mahagr** [comment]
You can disable file change checks, which may help a bit -- at least after the initial page load. But I don't think there's a real cure available at this moment, because of how (all) the pages are currently fetched in.

What I would do in the mean time is to split the site into 2-3 separate instances to keep the file count smaller. It would also be helpful if we could get hands on the data in order to figure out which are the bottlenecks as it would greatly help us on finding out the real issues.

**Comment by rhukster** [comment]
My idea is to have a CLI-based command that can pre-fetch and pre-cache all the pages. It would turn off the PHP timeout temporarily to do this, or do it in batches.  This way you can run it manually without relying on a page being hit to start the process.

**Comment by HOFB2015** [comment]
Hi,

So I disabled the file change checks via system.yaml and also increased php.ini timeout error to 45 sec's. This has helped in terms of each page takes about 35's to load. Gives me time to view changes and see what I have done and what needs changing.

I've looked at the pre-cache plugin and that would work once I get to the production state but as it creates over 250,000k folders its a pain to "clear cache" on development site where I am still making loads of small changes here and their. Not sure why the cache seems to want to create about 100 empty folders before placing the file after those 100 folders... something like 1/ab/54/3g/d/gd/fd/fd/fd/fd/fd/w/q/b/s/bs/b/sb/b/bd/b/r/4/3/2/b/s/3/5n//s/435/d/b/2/5/nb/d/filename.yaml.php

Will look at zipping up the site and stripping out the videos/images perhaps as its a few GB's in size.

**Comment by rhukster** [comment]
If you can get a zip file that is a decent size, I would love to download it and do some tests with it to try to optimize things as much as possible.

The cache mechanism is all handled by Doctrine Cache.  We really don't control it's internal folder storage structure, or anything like that.  However, the cache doesn't clear the old cache when it refreshes the cache, it just changes the cache key which effectively just changes the folder structure.  So the more changes you make, the more folders are created until you clear all.

Using a memory based cache (APC, XCache, Memcache) will help remove a bunch of those files making clearing cache much faster, will also speed things up a bit too.

**Comment by HOFB2015** [comment]
Zip sent to you and mahagr :+1:

**Comment by rhukster** [comment]
Ok some thoughts:

1) An update for Doctrine cache should limit the number of folders created for caching (coming in next release)
2) Using APC cache helps a bit in that it wont create all those folders and is slightly faster than file.
3) Using PHP 5.6 makes things a little faster too
4) The advanced_pagecache plugin i linked you too helps a TONNE after the initial cache process.
5) I noticed you had page-caching `false`, i set that to `true` so after initial cache, things are much faster

Overall on my local system, it takes 10seconds for the initial cache, 500ms for a cached page to load, and with the advanced_pagecache, basically 10ms.

I started to look at a populatecache CLI command, but ran into some issues because I have to try and emulate a full HTTP request, as that sets up the proper setup and configuration to ensure the cached version is 100% in sync with what would be cached from the web.  As this is quite complex, i've put it off for now.  Hitting the website once, does the exact same thing anyway.
